### PR TITLE
unlock with resource and id as argument

### PR DESCRIPTION
**Pull request by ar4hc**
simple solution for #83 

(and a typo...)

### CODE REVIEW

**Review by gtmanfred — CHANGES_REQUESTED** [review]
I think I would prefer to have a `get` function on the manager, that would allow for getting just one lock, similar to the `get_all_active_locks` function, and then leave `unlock()` unmodified.  This would allow you to get the lock object first, and pass it in.

**Comment by ar4hc** [comment]
sry, didn't look here in a while... :/ 
Where is that "get_all_active_locks" function...?

And as the unlock unpack the Lock object and uses the resource and identifier wouldn't that mean some extra steps...?
Creating a new Lock object just to have it  unpacked again in the next step looks a bit convoluted to me...

**Comment by ar4hc** [comment]
the `create_lock` method (more `fake_lock`...) would look like this:
```python
    async def create_lock(resource, lock_identifier, lock_timeout=10.0):

        lease_time = lock_timeout or self.internal_lock_timeout
        lock = Lock(self, resource, lock_identifier, lease_time, valid=True)
        return lock
```

perhaps add a check of the lock before with `is_locked`, but that's the gist of it.

i don't like it... :(

**Comment by gtmanfred** [comment]
We have other places where we make round trips to the database to make sure locks exist.  I would rather add that function, because i can see it be useful in other places.

You could also retrieve the ttl left on a lock to know when it is going to expire as opposed to setting it based on what is passed in.

Also, if we add another redis script that checks the identifier that is passed in, and returns the TTL, that would give us all the information we need to recreate the lock on one round trip.

**Comment by gtmanfred** [comment]
Something like this ```
    CHECK_LOCK_SCRIPT = """
    local identifier = redis.call('get', KEYS[1])
    if not identifier then
        return redis.error_reply('ERROR')
    if identifier == ARGV[1] then
        return redis.call("TTL", KEYS[1])
    else
        return redis.error_reply('ERROR')
    end"""
```

**Comment by ar4hc** [comment]
get_lock_ttl is a bit over complicated for that task, i think, but here you go... :)

File: tests/ut/test_redis.py [hunk 1]
```diff
@@ -475,3 +496,16 @@ async def test_clear_connections(self, mock_redis_two_instances):
         await redis.clear_connections()
 
         assert pool.close.called is False
+
+    @pytest.mark.asyncio
+    async def test_get_lock(self, mock_redis_two_instances, ):
+        redis, pool = mock_redis_two_instances
+
+        res = await redis.get_lock_ttl('resource', 'lock_id')
+        print(res)
```
Comment (gtmanfred) [inline tests/ut/test_redis.py:505]: looks like this is left behind.

File: tests/ut/test_redis.py [hunk 2]
```diff
@@ -475,3 +496,16 @@ async def test_clear_connections(self, mock_redis_two_instances):
...
+
+        script_sha1 = getattr(redis.instances[0], 'get_lock_ttl_script_sha1')
+
+        calls = [call(script_sha1, keys=['resource'], args=['lock_id'])]
+        pool.evalsha.assert_has_calls(calls)
+        # assert 0
```
Comment (gtmanfred) [inline tests/ut/test_redis.py:511]: cleanup

File: tests/ut/test_redis.py [hunk 3]
```diff
@@ -456,6 +476,7 @@ async def test_three_instances_combination(
 
         calls = [call(script_sha1, **call_args)] * 3
         pool.evalsha.assert_has_calls(calls)
+        # assert 0
```
Comment (gtmanfred) [inline tests/ut/test_redis.py:479]: cleanup

File: tests/ut/test_redis.py [hunk 4]
```diff
@@ -225,6 +226,21 @@ async def test_lock(self, fake_instance):
...
 
+    @pytest.mark.asyncio
+    async def test_get_lock_ttl(self, fake_instance, caplog):
+        caplog.set_level(logging.DEBUG)
+        instance = fake_instance
+        await instance.connect()
+        pool = instance._pool
+
+        ttl = await instance.get_lock_ttl('resource', 'lock_id')
+        print(ttl)
```
Comment (gtmanfred) [inline tests/ut/test_redis.py:237]: cleanup

File: tests/ut/test_redis.py [hunk 5]
```diff
@@ -179,7 +180,7 @@ def awaiter(self):
 
         assert instance._pool is None
         await instance.connect()
-        assert pool.execute.call_count == 2
+        assert pool.execute.call_count == len(self.script_names)
```
Comment (gtmanfred) [inline tests/ut/test_redis.py:183]: nice

File: aioredlock/algorithm.py [hunk 6]
```diff
@@ -10,6 +10,9 @@
 from aioredlock.lock import Lock
 from aioredlock.redis import Redis
 
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
```
Comment (gtmanfred) [inline aioredlock/algorithm.py:14]: cleanup

File: aioredlock/algorithm.py [hunk 7]
```diff
@@ -171,34 +174,45 @@ async def extend(self, lock, lock_timeout=None):
                 await self.unlock(lock)
             raise
 
-    async def unlock(self, lock):
+    async def unlock(self, resource_or_lock, lock_identifier=None):
```
Comment (gtmanfred) [inline aioredlock/algorithm.py:177]: I think that we should still just be passing in the lock here, now that you can get a recreated lock.  This was the whole point of adding `get_lock` to leave `unlock` unmodified.

**Comment by gtmanfred** [comment]
The rest of this looks good, just some cleanup. and removing the changes from the `unlock` function please.

**Review by gtmanfred — APPROVED** [review]
Looks great!

Thanks for making those changes.

**Comment by ar4hc** [comment]
:)

**Comment by gtmanfred** [comment]
You should be all set to go, this is released in v0.6.0, which i am pushing out now.

Thank you for the contribution!

**Pull request merged**